src/join_stream.rs

```diff
@@ -0,0 +1,82 @@
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use futures_core::Stream;
+
+/// A stream joining two or more streams.
+///
+/// This stream is returned by `join!`.
+#[derive(Debug)]
+pub struct JoinStream<L, R, T> {
+    left: L,
+    right: R,
+    _marker: PhantomData<T>,
```

Comment:
Is this the right `PhantomData`? We don't store `T` directly, so `fn() -> T` seems more appropriate. It also seems like we don't need this `T` at all. Here's how I imagine this can be implemented (with iterators): https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b86cc46552dac993c56a2dbde9db82ea

Reply:
Oh, good point! -- I've changed it!